src/main/scala/com/nec/spark/planning/VERewriteStrategy.scala

```diff
@@ -123,40 +121,35 @@ final case class VERewriteStrategy(options: VeRewriteStrategyOptions)
             if options.joinOnVe =>
           val functionName = s"join_${functionPrefix}"
 
-          val exchangeNameL = s"exchange_l_$functionPrefix"
-          val exchangeNameR = s"exchange_r_$functionPrefix"
-          val exchangeFunctionL = {
-            GroupingFunction.groupData(
-              data = inputsLeft.map(leftVector =>
-                DataDescription(
-                  veType = leftVector.veType,
-                  keyOrValue =
-                    if (genericJoiner.joins.flatMap(_.vecs).contains(leftVector)) KeyOrValue.Key
-                    else KeyOrValue.Value
-                )
-              ),
-              totalBuckets = HashExchangeBuckets
-            )
-          }
-          val exchangeFunctionR = {
-            GroupingFunction.groupData(
-              data = inputsRight.map(rightVector =>
-                DataDescription(
-                  veType = rightVector.veType,
-                  keyOrValue =
-                    if (genericJoiner.joins.flatMap(_.vecs).contains(rightVector)) KeyOrValue.Key
-                    else KeyOrValue.Value
-                )
-              ),
-              totalBuckets = HashExchangeBuckets
-            )
-          }
+          val exchangeFunctionL = GroupingFunction(
+            s"exchange_l_${functionPrefix}",
+            inputsLeft.map { vec =>
+              GroupingFunction.DataDescription(
+                vec.veType,
+                if (genericJoiner.joins.flatMap(_.vecs).contains(vec)) GroupingFunction.Key
+                else GroupingFunction.Value
+              )
+            },
+            HashExchangeBuckets
+          )
+
+          val exchangeFunctionR = GroupingFunction(
+            s"exchange_r_${functionPrefix}",
+            inputsRight.map { vec =>
+              GroupingFunction.DataDescription(
+                vec.veType,
+                if (genericJoiner.joins.flatMap(_.vecs).contains(vec)) GroupingFunction.Key
+                else GroupingFunction.Value
+              )
+            },
+            HashExchangeBuckets
+          )
 
           val code = CodeLines
             .from(
               CFunctionGeneration.KeyHeaders,
-              exchangeFunctionL.toCodeLines(exchangeNameL),
-              exchangeFunctionR.toCodeLines(exchangeNameR)
+              exchangeFunctionL.toCodeLines,
```

Comment:
Just curious what's your thinking here?

Reply:
The idea is to have `GroupingFunction` and similar classes be a full encoding that can render a `CFunction2`.   But since `CFunction2` is at a lower level abstraction than what it should be (e.g. doesn't have a `name: String` or `headers: Set[Header]`), `GroupingFunction` is responsible for directly emitting `CodeLines` instead.